Helper/Entity/Product/PriceManager.php

```diff
@@ -503,26 +507,108 @@ private function handleZeroDefaultPrice(
         $baseCurrencyCode,
         $min,
         $max,
-        Store $store
+        Store $store,
+        $subProducts,
+        $areCustomersGroupsEnabled,
+        $groups,
+        $specialPrice
     ) {
+        if ($min == $max) {
+            if (!$areCustomersGroupsEnabled && !empty($specialPrice)) {
+                $min = $max = $specialPrice[0];
+            } elseif ($areCustomersGroupsEnabled && !empty($specialPrice)) {
+                foreach ($groups as $group) {
+                    $groupId = (int)$group->getData('customer_group_id');
+
+                    $customData[$field][$currencyCode]['group_' . $groupId] = $specialPrice[$groupId];
+                    $groupSpecialPriceFormated = $this->priceCurrency->format(
+                        $specialPrice[$groupId],
+                        false,
+                        PriceCurrencyInterface::DEFAULT_PRECISION,
+                        $store,
+                        $currencyCode
+                    );
+                    $customData[$field][$currencyCode]['group_' . $groupId . '_formated'] = $groupSpecialPriceFormated;
+                }
+            }
+        }
+
         $customData[$field][$currencyCode]['default'] = $min;
 
+        $defaultOriginals = [];
```

Comment:
How about extracting code from here to line 593 to a separate method as it returns value on line 592? It'll make the code much readable and understandable.

Reply:
Yes good idea.